### DISCUSSION TITLE
Active/Active Keycloak cluster

### DISCUSSION THREAD

**Discussion by mirzlab** [keycloak/keycloak#30147] [Q&A]
Hello,

We are deploying Keycloak 24.x in AWS through AWS ALB (Application Load balancer) + AWS ECS (Docker Container) + RDS (PostgreSQL).

I have already read the discussions here : 
https://github.com/keycloak/keycloak/discussions/26440 
https://github.com/keycloak/keycloak/discussions/25269

But I wanted to create a new discussion hoping to get more insights regarding our specific architecture.

Our ECS nodes and RDS will be located in the same AWS region and availability zone (no cross-datacenters configurations).

For scalability during peak usage, we want to set up a cluster of Keycloak ECS nodes which are all active behind the load balancer.

Keycloak documentation states that only active/passive clusters are possible, but I read on few forums/discussions that active/active cluster can be achieved through external infinispan (but no "official" statement regarding this kind of configuration).

I am asking the community for some guidance if our goal for an active/active cluster is truly reachable within a same region :), and would be greatful for any advice.

Thank you very much for your help !

**Comment by ahus1** [comment]
If you don't have transparent networking between the regions and set it up with an external Infinispan as described in the Keycloak HA guide, you'll only be able to do A/P for now. 

At the same time, if you have transparent networking and deploy your Keycloak cluster without an external Infinispan and would connect all your Keycloaks directly with JGroups, that would technically work. At the same time you would see more traffic and increased delays between  the regions because requests would go back-and-forth. Depending on the number of owners you configure and the number of Keycloak instances you deploy, you will see an incoming read-only request read session information from a different site, and a write request updating the primary owner in the other site which is then updating the secondary owner in a third site. The networking between the sites can also become a problem when it is becomes unreliable and it would lead to timeouts and errors, and you'll need to find a way to re-synchronize sites when that happens. 

Those are the reasons why this hasn't (yet) been documented in Keycloak's documentation.
